File: cacao/ui/components/data/tree_viewer/tree_viewer.py
from typing import Any, Dict, Optional
from ...base import Component

class TreeViewer(Component):
    """
    A component to render collapsible tree structures.

    Props:
      - id: unique component ID
      - data: the object to display as a tree
      - expand_all: whether to start fully expanded
      - theme: "light" or "dark"
      - on_node_click: optional event name to fire when a key is clicked
    """
    def __init__(
        self,
        id: str,
        data: Dict[str, Any],
        expand_all: bool = False,
        theme: str = "light",
        on_node_click: Optional[str] = None,
        **kwargs
    ):
        super().__init__(**kwargs)
        self.id = id
        self.data = data
        self.expand_all = expand_all
        self.theme = theme
        self.on_node_click = on_node_click

    def validate(self) -> bool:
        return isinstance(self.data, (dict, list))

    def to_dict(self) -> Dict[str, Any]:
        props = {
            "id": self.id,
            "data": self.data,
            "expand_all": self.expand_all,
            "theme": self.theme,
        }
        if self.on_node_click:
            props["on_node_click"] = self.on_node_click

        return {
            "type": "tree_viewer",
            "props": props,
            **self.get_base_props()
        }

File: cacao/ui/components/data/tree_viewer/tree_viewer.js
(component) => {
    console.log("[CacaoCore] Rendering tree_viewer component:", component);
    const tree = document.createElement('div');
    tree.className = 'tree_viewer';
    
    // Extract props
    const expandAll = component.props.expand_all || false;
    const onNodeClick = component.props.on_node_click;
    const data = component.props.data;
    const theme = component.props.theme || 'light';
    
    // Apply theme class
    tree.classList.add(`theme-${theme}`);
    
    // Set ID if provided
    if (component.props.id) {
        tree.id = component.props.id;
    }

    function renderNode(key, value, parent) {
        const node = document.createElement('div');
        node.className = 'tree-node';
        const isObject = value !== null && typeof value === 'object';

        // toggle handle
        const toggle = document.createElement('span');
        toggle.className = 'tree-expand-toggle';
        toggle.textContent = isObject ? (expandAll ? '▼' : '▶') : '';
        node.appendChild(toggle);

        // key
        const keySpan = document.createElement('span');
        keySpan.className = 'tree-key';
        keySpan.textContent = key;
        node.appendChild(keySpan);

        node.appendChild(document.createTextNode(':'));

        // primitive value
        if (!isObject) {
            const val = document.createElement('span');
            val.className = 'tree-value';
            val.textContent = JSON.stringify(value);
            node.appendChild(val);
        }

        // children
        if (isObject) {
            const childrenWrapper = document.createElement('div');
            childrenWrapper.className = 'tree-children';
            if (!expandAll) {
                childrenWrapper.style.display = 'none';
                node.classList.add('collapsed');
            }
            Object.entries(value).forEach(([k, v]) =>
                renderNode(k, v, childrenWrapper)
            );
            node.appendChild(childrenWrapper);

            toggle.addEventListener('click', () => {
                const collapsed = node.classList.toggle('collapsed');
                childrenWrapper.style.display = collapsed ? 'none' : 'block';
                toggle.textContent = collapsed ? '▶' : '▼';
            });
        }

        // optional click event
        if (onNodeClick) {
            keySpan.style.cursor = 'pointer';
            keySpan.addEventListener('click', () => {
                const evt = new CustomEvent(onNodeClick, { detail: { key } });
                tree.dispatchEvent(evt);
            });
        }

        parent.appendChild(node);
    }

    if (typeof data === 'object' && data !== null) {
        Object.entries(data).forEach(([k, v]) => renderNode(k, v, tree));
    }
    
    return tree;
}
